src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineRewriter.cs

```diff
@@ -325,5 +325,33 @@ protected SynthesizedImplementationMethod OpenMoveNextMethodImplementation(Metho
             F.CurrentFunction = result;
             return result;
         }
+
+        /// <summary>
+        /// Produce Environment.CurrentManagedThreadId if available, otherwise CurrentThread.ManagedThreadId
+        /// </summary>
+        protected BoundExpression MakeCurrentThreadId()
+        {
+            Debug.Assert(CanGetThreadId());
+            var currentManagedThreadIdProperty = (PropertySymbol)F.WellKnownMember(WellKnownMember.System_Environment__CurrentManagedThreadId, isOptional: true);
+            if ((object)currentManagedThreadIdProperty != null)
+            {
+                MethodSymbol currentManagedThreadIdMethod = currentManagedThreadIdProperty.GetMethod;
+                if ((object)currentManagedThreadIdMethod != null)
+                {
+                    return F.Call(null, currentManagedThreadIdMethod);
+                }
+            }
+
+            return F.Property(F.Property(WellKnownMember.System_Threading_Thread__CurrentThread), WellKnownMember.System_Threading_Thread__ManagedThreadId);
+        }
+
+        /// <summary>
+        /// Returns true if either Thread.ManagedThreadId or Environment.CurrentManagedThreadId are available
+        /// </summary>
+        protected bool CanGetThreadId()
```

Comment:
Doesn't this method need to verify that the `GetMethod` exists for `System_Environment__CurrentManagedThreadId`? The current invariant guaranteed here doesn't line up with what is needed in `MakeCurrenThreadId`

Reply:
This method is just a refactoring from IteratorRewriter (extracted for later re-use)

---
In reply to: [221787577](https://github.com/dotnet/roslyn/pull/30105#discussion_r221787577) [](ancestors = 221787577)